### PR TITLE
$this->mustache() w/ partials, git-sync fix

### PR DESCRIPTION
**Pull request by will123195**
`$this->mustache($mustache, $data, $partials);`

$mustache - can be mustache markup or a relative filename
$data - can be an object with properties/methods or array of values/functions
$partials - path to partials or array of partials (name => markup) or (name => filename)
## 

also fixed git-sync issue

also patched some security holes and added `safe_exec()`

### CODE REVIEW

File: lib/Sky/Page.php [hunk 1]
```diff
@@ -507,6 +507,32 @@ private function get_template_contents($template_name, $template_area)
     }
 
     /**
+     * Renders a mustache template using the specified data
+     * @param string $mustache the mustache filename (relative to php file or codebase)
+            OR mustache template markup string containing at least one {{variable}}
+     * @param mixed $data object or array of properties and/or functions
+     * @return string
+     */
+    public function mustache($mustache, $data)
```
Comment (stanistan) [inline lib/Sky/Page.php]: What about partials?

Comment (will123195) [inline lib/Sky/Page.php]: word.. i'll look into that

**Comment by will123195** [comment]
ready

File: lib/Sky/Mustache.php [hunk 2]
```diff
@@ -0,0 +1,133 @@
+<?php
+
+namespace Sky;
+
+class Mustache {
```
Comment (stanistan) [inline lib/Sky/Mustache.php]: opening brace

File: lib/Sky/Mustache.php [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    /**
+     * Sky Mustache Constructor
+     * @param string $mustache mustache filename (relative to calling php file or codebase)
+            OR mustache markup string containing at least one {{tag}}
+     * @param mixed $data object with properties/methods or array of values/functions
+     * @param mixed $partials path to partials or array of name => filename/markup
+     * @param string $path path to check for the main markup file (if applicable)
+     * @return string
+     */
+    public function __construct($mustache, $data, $partials=null, $path=null)
```
Comment (stanistan) [inline lib/Sky/Mustache.php]: spaces in default args

File: lib/Sky/Mustache.php [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        }
+
+        $this->markup = $markup;
+        $this->data = $data;
+        $this->partials = $this->getPartials($markup, $paths, $partials);
+
+    }
+
+    /**
+     * Renders the mustache
```
Comment (stanistan) [inline lib/Sky/Mustache.php:62]: @return string

File: lib/Sky/Mustache.php [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+    /**
+     * Recursively get the markup for each partial needed. Append any newly found partials
+     * to the known $partials array.
+     * @param string $markup the mustache doc to check for partials
+     * @param array $paths the list of paths to check where each partial may exist
+     * @param array $partials the array of partials already known
+     * @return array
+     */
+    private function getPartials($markup, $paths=null, $partials=null)
```
Comment (stanistan) [inline lib/Sky/Mustache.php]: should default values for $paths/$partials here be `array()` ?

File: lib/Sky/Mustache.php [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+            if (!$partials[$name]) {
+                foreach ($paths as $path) {
+                    $partials[$name] = $this->getMarkup($name, $path);
+                    if ($partials[$name]) {
+                        $markup = $partials[$name];
+                        $partials = $this->getPartials($markup, $paths, $partials);
+                        break;
+                    }
+                }
+                if (!$partials[$name]) throw new \Exception(
```
Comment (stanistan) [inline lib/Sky/Mustache.php]: weird formatting

File: lib/core/functions.inc.php [hunk 7]
```diff
@@ -1828,6 +1796,23 @@ function aql2array($param1, $param2 = null) {
 }
 
 
+/**
+ * Escapes the given command and executes it on the command line
+ * @param mixed $command command string or array of command strings
+ * @param array $output
+ * @return array
+ */
+function safe_exec($command, &$output=null) {
```
Comment (stanistan) [inline lib/core/functions.inc.php]: opening brace.

Why is `output` a reference? and why is it set to null and not `array`

File: lib/dev/get-codebase.inc.php [hunk 8]
```diff
@@ -44,9 +51,12 @@ function getCodeBase($codebase_path, $codebase)
...
         mkdir($branch_path, 0777, true);
 
-        $cmd = "cd $branch_path; git clone -b $branch git@github.com:$user/$repository.git . 2>{$codebase_path}git.log;";
+        $cmds = array(
+            "cd $branch_path",
+            "git clone -b $branch git@github.com:$user/$repository.git . 2>{$codebase_path}git.log"
+        );
 
-        echo exec($cmd);
+        echo safe_exec($cmds);
```
Comment (stanistan) [inline lib/dev/get-codebase.inc.php:59]: this is weird, doc says this function returns an array

File: lib/Sky/Page.php [hunk 9]
```diff
@@ -507,6 +507,28 @@ private function get_template_contents($template_name, $template_area)
...
+            OR mustache template markup string containing at least one {{tag}}
+     * @param mixed $data object or array of properties and/or functions
+     * @param mixed $partials path to partials or array of partial => filename
+     * @return string
+     */
+    public function mustache($mustache, $data, $partials=null)
+    {
+        // get the path of the php file calling this method in case we need to look for
+        // partials with a relative file name
+        // TODO: php 5.4, use second parameter to set backtrace limit = 1
```
Comment (stanistan) [inline lib/Sky/Page.php]: remove trace

**Comment by will123195** [comment]
fixed

**Pull request merged**